### PR TITLE
Don't trigger `filter_map_identity` with an iterator from an empty array

### LINKED ISSUE #12653: compilation failure after applying lint `filter_map_identity`
### Summary

title, the following code depends on `filter_map` to tell rustc that the type is Option<_>, which is removed causing a compilation error

### Reproducer

I tried this code:

```rust
fn iter() -> impl Iterator<Item = u8> {
    [].into_iter().filter_map(|x| x)
}
```

I expected to see this happen:
no lint (see #9377) or a working one

Instead, this happened:
command: `cargo clippy --fix --allow-no-vcs`
```
    Checking rust-playground v0.1.0 (/home/antoni/code/rust-playground)
warning: failed to automatically apply fixes suggested by rustc to crate `rust_playground`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
 --> src/main.rs:2:8
  |
2 |     [].into_iter().flatten()
  |     -- ^^^^^^^^^
  |     |
  |     type must be known at this point
  |
help: try using a fully qualified path to specify the expected types
  |
2 |     <[_; 0] as std::iter::IntoIterator>::into_iter([]).flatten()
  |     +++++++++++++++++++++++++++++++++++++++++++++++  ~

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: use of `filter_map` with an identity function
 --> src/main.rs:2:20
  |
2 |     [].into_iter().filter_map(|x| x)
  |                    ^^^^^^^^^^^^^^^^^ help: try: `flatten()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#filter_map_identity
  = note: `#[warn(clippy::filter_map_identity)]` on by default

warning: `rust-playground` (bin "rust-playground") generated 1 warning (run `cargo clippy --fix --bin "rust-playground"` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `rust_playground`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
 --> src/main.rs:2:8
  |
2 |     [].into_iter().flatten()
  |     -- ^^^^^^^^^
  |     |
  |     type must be known at this point
  |
help: try using a fully qualified path to specify the expected types
  |
2 |     <[_; 0] as std::iter::IntoIterator>::into_iter([]).flatten()
  |     +++++++++++++++++++++++++++++++++++++++++++++++  ~

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: `rust-playground` (bin "rust-playground" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.27s
```

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

### PR DESCRIPTION
**Pull request by lapla-cogito**
fix #12653

changelog: [`filter_map_identity`]: don't lint for creating an iterator from an empty array

### CODE REVIEW

File: clippy_lints/src/methods/filter_map_identity.rs [hunk 1]
```diff
@@ -21,6 +21,14 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &hir::Expr<'_>, filter_map_arg:
     if is_trait_method(cx, expr, sym::Iterator)
         && let Some(applicability) = is_identity(cx, filter_map_arg)
     {
+        if let Some((_, recv, _, _, _)) = method_call(expr)
+            && let Some((_, recv, _, _, _)) = method_call(recv)
+            && let hir::ExprKind::Array(arr) = recv.kind
```
Comment (Jarcho) [inline clippy_lints/src/methods/filter_map_identity.rs:26]: Use `let ExprKind::MethodCall(recv, ..) = expr.kind` rather than `method_call`.

**Review by Jarcho — CHANGES_REQUESTED** [review]

File: clippy_lints/src/methods/filter_map_identity.rs (hunk 1 above)
Comment (lapla-cogito) [reply clippy_lints/src/methods/filter_map_identity.rs]: Fixed in https://github.com/rust-lang/rust-clippy/pull/13826/commits/d3d24e70ca59b2404510356e3be857045eab707b.

**Comment by samueltardieu** [comment]
Couldn't we suggest using `{std,core}::iter::empty::<T>()` instead?

**Review by Jarcho — APPROVED** [review]
Thank you.

**Comment by Jarcho** [comment]
`iter::empty` would be close enough to the same thing (different type, but close given the array length is part of the type). Could arguably be a different lint though that catches different ways of writing an empty iterator.

**Pull request merged**